cmake/compile_tests/cplusplus17.cpp

```diff
@@ -0,0 +1,8 @@
+#include <type_traits>
+
+int main() {
+  // _v versions of type traits were added in C++17
+  constexpr bool same = std::is_same_v<double, int>;
```

Comment:
Is it salient that this is testing a standard library feature, rather than a compiler feature? Variable templates themselves were part of C++14, so a standard library implementation could provide an implementation for this that would compile under C++14.

Reply:
Previously, we had `std::remove_cv_t<int> i = 0;` for C++14.
I am happy to hear other suggestions. What about `if constexpr`?

Reply:
As an edge example, libstdc++ as shipped with GCC 7.1 would pass this, though we're planning to set at least 7.3 as the minimum per https://github.com/kokkos/kokkos/issues/5285

Reply:
Given that `if constexpr` is probably one of the most desirable headline features for us, that seems quite suitable as a target.

Reply:
I added `if constexpr` but also kept `std::is_same_v` since the test case was initially designed to detect a working toolchain (with recent enough gcc headers), see https://github.com/kokkos/kokkos/pull/3809 and https://github.com/kokkos/kokkos/issues/3789.